### PR TITLE
[MPS] Migrate gelu and gelu_backward to metal

### PR DESCRIPTION
**Pull request by jhavukainen**
Partially answers the concerns in issue #181213 migrating the gelu and gelu_backward including the tanh approximation to metal kernels. Below will be some implementation considerations as well as benchmarking for the implementation.

For bf16/fp16 registrations for gelu_tanh use metal::fast::tanh as an experimental sweep concluded that numerically the discrepancy between the fast and precise modes used will be confined to reasonably low absolute errors. Note the intermediate computations are still done in fp32 but tanh uses a faster approximation. Attaching the summary plots for both dtypes in backward and forward configurations displaying both the absolute error and the corresponding discrepancy in ULPs for the storage dtype is scanned between [-10.0, 10.0]. The region where the function approaches zero displays the largest discrepancy in ULPs due to the values being clustered more tightly together near zero especially for fp16. However the absolute error between the modes in that region stays around 10^-5 so the overall effect from the approximation seems well behaved by the experiment.

gelu_tanh_fast_accuracy

gelu_tanh_backwards_fast_accuracy

### Benchmarks

The fresh benchmarks on a M5 Max. Extended the shapes with a couple of example shapes from realistic workloads to get a more comprehensive view. The benchmarking script below and a graph summarizing the delta after measured both with the per-op sync as well as the tight encode loop with a single sync (per-batch in the script). As was discussed below in the thread this affects the performance of MPSGraph as commitAndContinue gives it some freedom to optimize how the computations are dispatched.

```
import argparse
import statistics
import time
import torch
import torch.nn.functional as F

DTYPE_TAG = {torch.float32: "f32", torch.float16: "f16", torch.bfloat16: "bf16"}

BERT_SHAPES = [
    (1, 128, 4096),
    (1, 1024, 4096),
    (1, 4096, 4096),
    (4, 128, 4096),
    (4, 1024, 4096),
    (16, 128, 4096),
    (1, 8192, 8192),
    (4, 8192, 8192),
]

LLM_SHAPES = [
    (1, 2048, 11008),
    (1, 4096, 11008),
    (4, 2048, 13824),
]

MODES = ("none", "tanh")

parser = argparse.ArgumentParser()
parser.add_argument("--sync", choices=("per-iter", "per-batch"), default="per-iter")
parser.add_argument("--warmup", type=int, default=20)
parser.add_argument("--iters", type=int, default=300)
parser.add_argument("--samples", type=int, default=10)
args = parser.parse_args()


if args.sync == "per-iter":
    def time_op(fn):
        torch.mps.synchronize()
        t0 = time.perf_counter()
        for _ in range(args.iters):
            fn()
            torch.mps.synchronize()
        return (time.perf_counter() - t0) / args.iters * 1e6
else:
    def time_op(fn):
        torch.mps.synchronize()
        t0 = time.perf_counter()
        for _ in range(args.iters):
            fn()
        torch.mps.synchronize()
        return (time.perf_counter() - t0) / args.iters * 1e6


cases = []
for s in BERT_SHAPES:
    for mode in MODES:
        cases.append((s, torch.float32, mode))
for s in LLM_SHAPES:
    for dtype in (torch.float16, torch.bfloat16, torch.float32):
        for mode in MODES:
            cases.append((s, dtype, mode))


ops = []
for shape, dtype, mode in cases:
    x = torch.randn(*shape, device="mps", dtype=dtype)
    fwd = lambda x=x, m=mode: F.gelu(x, approximate=m)
    xg = torch.randn(*shape, device="mps", dtype=dtype, requires_grad=True)
    g = torch.ones_like(xg)

    def bwd(xg=xg, g=g, m=mode):
        xg.grad = None
        F.gelu(xg, approximate=m).backward(g)

    tag = f"{mode[0]}_{DTYPE_TAG[dtype]}"
    ops.append((shape, tag, fwd, bwd))

for _, _, fwd, bwd in ops:
    for _ in range(args.warmup):
        fwd()
    for _ in range(args.warmup):
        bwd()

fwd_samples = {(s, t): [] for s, t, _, _ in ops}
bwd_samples = {(s, t): [] for s, t, _, _ in ops}
for _ in range(args.samples):
    for s, t, fwd, bwd in ops:
        fwd_samples[(s, t)].append(time_op(fwd))
        bwd_samples[(s, t)].append(time_op(bwd))

print(f"{'shape':<22} {'tag':<8} {'fwd_us':>10} {'fwd_sd':>8} {'bwd_us':>10} {'bwd_sd':>8}",
      flush=True)
for shape, dtype, mode in cases:
    tag = f"{mode[0]}_{DTYPE_TAG[dtype]}"
    f = fwd_samples[(shape, tag)]
    b = bwd_samples[(shape, tag)]
    print(f"{str(shape):<22} {tag:<8} "
          f"{statistics.mean(f):>10.1f} {statistics.stdev(f):>8.1f} "
          f"{statistics.mean(b):>10.1f} {statistics.stdev(b):>8.1f}",
          flush=True)
```

gelu_perf_bars_sync_nosync

To summarize the results the metal op implementation seems good enough here after the recent changes added for the vectorized versions being registered for both forward and backward calls. As discussed below the commitAndContinue mechanism makes comparing the methods a bit difficult due to metal kernels and MPSGraph handling it differently. So the "true performance" depends a bit on the context the op is used and is likely going to be somewhere between the two sets of measurements (synced vs. no sync) presented here. But overall the benefit seems to be towards the metal kernel with maybe the backward pass no sync measurements being more on the fence. However especially smaller shapes the metal kernel shows benefits. Also for the LLM inspired shapes there seems to a trend towards the metal kernel being slightly but consistently better than the pre-existing implementation.

### CODE REVIEW

**Comment by Isalia20** [comment]
I tried porting gelu but had quite hard performance regressions on float16 tensors, so worth looking into it I think

**Comment by jhavukainen** [comment]
Thanks that's a good point @Isalia20. I'll need to also re-check this on an M5 device since they might have improved performance on the half types depending on the kernel...

**Review by malfet — APPROVED** [review]

**Comment by jhavukainen** [comment]
I did some additional testing on perf based on the comment about the earlier effort. In summary the metal current implementation should be fine based on both M2 and M5 measurement. One issue that may show MPSGraph being significantly faster is if the op is tested in a loop such as 

```
for n in range(iter):
  _ = op
torch.mps.synchronize()
```

to dispatch a batch of encoded work. Which one might assume leads to similar measurements regardless of having implemented the op through MPSGraph or metal kernel since we have the "commit and continue" type of option for dispatch methods. However the difference is that when MPSGraph is given `_executionDescriptor.enableCommitAndContinue = _enableCommitAndContinue;` flag it has a bit more autonomy and it can decide to commit the buffer whenever deemed optimal to kickstart part of the encoded batch of ops. The metal kernel queue will in contrast keep accumulating unless some other op or memory pressure forces the commit. So to get comparable measurements one can either do synchronize after each op which makes MPSGraph look bad since you'll get the overhead per op even though the commitAndContinue tries to ameliorate just that. Or try to figure out some size of "batching" the op calls and call the synchronize that might somewhat overlap the MPSGraph one. Or try to benchmark it in a more realistic use case context like as a part of a transformer layer using the op or similar but that'll of course add a bunch of other noise and make it harder to distinguish the signal.

But that being said the current implementation looks ok in realistic usage context as well as doing the per op synchronize (or by guessing a good batching number for the synchronize calls).

There is also a real perf improvement that could be made but it'll be a small tradeoff between accuracy and speed: The current formulation is similar to what MLX uses and it reaches a pretty good accuracy, they quote ~0.99 ULP max deviation for FP32 presumably near 1.0. But the cost is that the kernel has branching in the erf computation which can cause warp divergences. One can use a branchless approximation which has a max error of  1.5 * 10^-7 should correspond to 1.26 ULPs near 1.0 but it gets rid of the divergence making it measurably faster at least on op level by some ~5+% especially for small shapes. It still does pass all the accuracy comparisons in the default testing for gelu fwd and backwards so if those are considered enough it might make sense to use this approach. On the other hand sticking to the same algorithm the other frameworks use at least gives us a guarantee of having similar accuracy and the small improvement to gelu seems to stay below the noise floor when benchmarking some more realistic use case like a transformer layer.

```
  gelu_branchless_erf(u):                                                                                                                                                                              
      # A&S 7.1.26 fixed coefficients                                                                                                                                                                  
      p  = 0.3275911                                                                                                                                                                                   
      a1 = 0.254829592                                                                                                                                                                                 
      a2 = -0.284496736                                                                                                                                                                                
      a3 = 1.421413741                                                                                                                                                                                 
      a4 = -1.453152027                                                                                                                                                                                
      a5 = 1.061405429                                                                                                                                                                                 
                                                                                                                                                                                                       
      t    = 1 / (1 + p * |u|)
      eu   = exp(-u * u)
      poly = (((( a5*t + a4 )*t + a3 )*t + a2 )*t + a1 )*t                         
      return copysign(1 - eu * poly, u)
```

This skips the long and short range polynomial split in the existing erf implementation. If this sounds like something we should do I can amend the PR with the change. But that being said: with the context of the differences in measuring at op level the speedup of metal kernel vs. MSPGraph at the top I would say the current implementation seems to perform on-par to the graph path in the realistic use and the per op perf testing and there is also value in following the existing "battle tested" implementations.

@Isalia20: Do you happen to have any specific test you recall showing the regression? I'd like to confirm on that use case as well if its still around just to be sure I've covered all the bases since it's such a clear cut case to try to encompass all possible usages where one implementation might show improvement over the other.

**Comment by Isalia20** [comment]
Did you do a few warmup runs before entering the benchmark with:
```
for n in range(iter):
  _ = op
torch.mps.synchronize()
```

something like:
```
for _ in range(5):
    _ = op
torch.mps.synchronize()

for n in range(iter):
  _ = op
torch.mps.synchronize()
```

The first PSO call compiles the Metal shader and builds the pipeline state, which gets cached after that. Without warmup, that one-time compile cost gets counted in your timing and skews the numbers.

As for which shapes, it was usually the fp16 large enough tensors on M5 Pro

**Comment by jhavukainen** [comment]
@Isalia20 Thanks,  yeah sorry should have mentioned the warmups are there but yes they are necessary. Both MPSGraph/MPS and metal kernels take the hit on the first call to the op before it's cached so they need to be excluded.

Ok I think the issue here is that I don't have a Pro version of the M5. With the base version GPU the low number of cores are pretty easy to keep saturated I think so the discrepancy between a well made kernel and a naive one tends to show up with the larger GPU count variants using larger workloads where one might need some finesse to best utilize the resources.

Let me see if I can access a higher tier device for a bit to scan the larger values. I'll pick https://github.com/pytorch/pytorch/pull/182155 as well since that should close the perf gap somewhat.

**Comment by Isalia20** [comment]
I can test it out for you on M5 Pro. Will running the benchmark script in the PR description suffice?

**Comment by malfet** [comment]
@Isalia20 perf regressions are much smaller if one rebases this change on top of https://github.com/pytorch/pytorch/pull/182155 (which adds ILP trick for backward)

File: aten/src/ATen/native/mps/kernels/ActivationKernel.metal [hunk 1]
```diff
@@ -227,3 +227,66 @@ struct silu_backward_functor {
...
+
+struct gelu_tanh_backward_functor {
+  template <typename T>
+  inline T operator()(const T grad, const T self) {
+    const float xf = float(self);
+    constexpr float kBeta = M_SQRT2_F * M_2_SQRTPI_F * 0.5f;
+    constexpr float kKappa = 0.044715f;
+    const float x_sq = xf * xf;
+    const float inner = kBeta * (xf + kKappa * xf * x_sq);
+    const float th = ::metal::tanh(inner);
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ActivationKernel.metal]: I'm debating with myself if this is an acceptable compromise to match the perf, especially for lower precision dtypes like float16/bfloat16
```suggestion
    const float th = ::metal::fast::tanh(inner);
```

Comment (jhavukainen) [reply aten/src/ATen/native/mps/kernels/ActivationKernel.metal]: I ran a sweep to get experimental bounds for the diff between fast and precise modes between inputs in the range of [-10.0, 10.0] both in absolute error as well as ULPs for fp16 and bf16. I'll add the plots for both forward and backward passes of gelu_tanh below but based on the results I would be confident in using the fast mode for fp16/bf16: In terms of ULPs there is a large discrepancy near where the function approaches zero but before it snaps into exactly zero especially for fp16 where the distance between ULPs becomes very small. However in absolute terms the discrepancy between the fast and precise modes stays well bounded below 10^-5 in that region so the effect on real computations in fp16 is should not sway the results significantly.

Image

Image

I'll rebase, switch to using the fast implementations and rerun the benchmark to confirm the perf.

Comment (malfet) [reply aten/src/ATen/native/mps/kernels/ActivationKernel.metal]: But make sure to include those graphs considerations in PR description

File: aten/src/ATen/native/mps/kernels/ActivationKernel.metal [hunk 2]
```diff
@@ -227,3 +227,75 @@ struct silu_backward_functor {
 REGISTER_BINARY_OP(silu_backward, float, float);
 REGISTER_BINARY_OP(silu_backward, half, half);
 REGISTER_BINARY_OP(silu_backward, bfloat, bfloat);
+
+template <typename T>
+static inline float gelu_dispatch_tanh(float x) {
+  if constexpr (::metal::is_same_v<T, float>) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ActivationKernel.metal]: Nit
```suggestion
  if IF_CONSTEXPR (::metal::is_same_v<T, float>) {
```

File: aten/src/ATen/native/mps/kernels/ActivationKernel.metal [hunk 3]
```diff
@@ -227,3 +227,75 @@ struct silu_backward_functor {
 REGISTER_BINARY_OP(silu_backward, float, float);
 REGISTER_BINARY_OP(silu_backward, half, half);
 REGISTER_BINARY_OP(silu_backward, bfloat, bfloat);
+
+template <typename T>
+static inline float gelu_dispatch_tanh(float x) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ActivationKernel.metal:232]:
```suggestion
static inline T gelu_dispatch_tanh(T x) {
```

Comment (jhavukainen) [reply aten/src/ATen/native/mps/kernels/ActivationKernel.metal:232]: Here the slightly unconventional use of template type with a locked in signature type is actually intentional: We only need it for the choosing the precise or the fast path for the tanh computation when the registered storage type is a half type and the precision loss from the fast path is insignificant in relation. But the input and the output will be float regardless as the rest of the kernel uses high precision for the intermediates and just downcast the result for storage. For example in the backward pass:

```
    const float inner = kBeta * (xf + kKappa * xf * x_sq);
    const float th = gelu_dispatch_tanh<T>(inner);
    const float dth = 1.0f - th * th;
```
we would just immediately upcast it back to float regardless. Similarly in forward pass.

We could maybe change `typename T` to `typename StorageT` or something similar to clarify the templating here?
